File: new/150_evaluate_reverse_polish_notation_recursion_dfs.cpp
class Solution {
public:
    // idea: recursion (dfs)
    // pop back -> if operator, traverse right subtree, traverse left subtree ....
    // if operand -> go left to right, and do the correct operation with left and right in that order. 
    int dfs(vector<string>& tokens) {
        string token = tokens.back();
        tokens.pop_back(); // in c++ this won't return the popped value, we use back() for this. 

        if (token != "+" && token != "-" && token != "*" && token != "/") {
            return stoi(token);
        }

        int right = dfs(tokens);
        int left = dfs(tokens); // this will give the left and right values for the corresponding
        // operator in the current stack. 

        if (token == "+") {
            return left + right;
        } else if (token == "-") {
            return left - right;
        } else if (token == "*") {
            return left * right;
        } else if (token == "/") {
            return left / right;
        }

        return -1;
    }

    int evalRPN(vector<string>& tokens) {
        return dfs(tokens);
    }

    // time: O(n)
    // space: O(n)
};

File: new/python/150_evaluate_reverse_polish_notation_recursion_dfs.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # idea: recursion (dfs)
        # ["4", "5", "2", "/", "+"]

        # We always pop from the RIGHT side (end of the list).
        # If the popped value is a NUMBER → return it immediately.
        # If the popped value is an OPERATOR → recursively evaluate:
        #       first its RIGHT operand,
        #       then its LEFT operand.

        # Walkthrough:

        # 1st pop: "+"   → operator → evaluate right subtree
        # 2nd pop: "/"   → operator → evaluate right subtree
        # 3rd pop: "2"   → number   → return 2  (this becomes the RIGHT operand of "/")
        # 4th pop: "5"   → number   → return 5  (this becomes the LEFT operand of "/")

        # Now evaluate this subtree:  5 / 2

        # Back to "+":
        # 5th pop: "4"   → number   → return 4  (this becomes the LEFT operand of "+")

        # Final expression:  4 + (5 / 2)
        def dfs():
            token = tokens.pop()
            if token not in "+-*/":
                return int(token)

            right = dfs()
            left = dfs()

            if token == "+":
                return left + right
            if token == "-":
                return left - right
            if token == "*":
                return left * right
            if token == "/":
                return int(left / right)

        return dfs()
        # time: O(n)
        # space: O(n)
